packages/p3/src/main/ts/sinap/decortors/index.ts

```diff
@@ -8,14 +9,86 @@ export type TSinapSuggest = {
   locale: string
 }
 
+export class SinapSuggestRequest implements TSinapSuggest {
+
+  query: string
+  fields?: Record<string, any>
+  locale: string
+
+  constructor(request: TSinapSuggest) {
+    this.fields = request.fields
+    this.locale = request.locale
+    this.query = request.query
+  }
+
+}
+
 export type TSinapContext = {
   fields?: Record<string, any>
   locale: string
 }
 
+export class SinapContextRequest implements TSinapContext {
+
+  fields?: Record<string, any>
+  locale: string
+
+  constructor(request: TSinapContext) {
+    this.fields = request.fields
+    this.locale = request.locale
+  }
+
+}
+
+export class SinapSuggestResponse<T = object> extends Array<T> {}
+
+export class SinapContextResponse {
+
+  [key: string]: any
+
+}
+
+const returnTypeMap = {
+  SinapSuggestResponse: 'SinapSuggest',
+  SinapContextResponse: 'SinapContext',
+}
+
+const paramTypeMap = {
+  SinapSuggestRequest: 'SinapSuggest',
+  SinapContextRequest: 'SinapContext',
+}
+
+const decoratorCompatibilityCheck = (
+  proto: IProto,
+  propName: string | undefined,
+  method: string,
+) => {
+  const returnType = Reflect.getMetadata('design:returntype', proto, propName + '')?.name
+  // @ts-ignore
+  if (returnType && returnTypeMap[returnType] !== method) {
+    throw new Error(`Cannot return ${returnType} in ${method}`)
+  }
+
+  const sinapParam = Reflect.getMetadata('design:paramtypes', proto, propName + '')
+    // @ts-ignore
+    .filter((el: any) => paramTypeMap[el.name] !== undefined)
```

Comment:
This filter finds implementations, but there's a need for decorators. They all are ctor in meta, once @params decorators was applied.